### PR TITLE
Report liveness from various loops.

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

In corner cases, the calculation graph and internal dataplane can take a long time to process some updates. Report liveness from their various loops.

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: calc/async_calc_graph.go [hunk 1]
```diff
@@ -149,6 +150,9 @@ func (acg *AsyncCalcGraph) loop() {
 					typeName := reflect.TypeOf(upd.Key).Name()
 					count := countUpdatesProcessed.WithLabelValues(typeName)
 					count.Inc()
+					if time.Since(acg.lastReport) > healthInterval {
```
Comment (neiljerram) [inline calc/async_calc_graph.go]: Is the elapsed time check really useful here?  It would be simpler to call `acg.reportHealth` unconditionally, so should we just do that?

File: dataplane/linux/int_dataplane.go [hunk 2]
```diff
@@ -268,6 +269,7 @@ func NewIntDataplaneDriver(config Config) *InternalDataplane {
 		LockProbeInterval:     config.IptablesLockProbeInterval,
 		BackendMode:           backendMode,
 		LookPathOverride:      config.LookPathOverride,
+		OnStillAlive:          dp.reportHealth,
```
Comment (neiljerram) [inline dataplane/linux/int_dataplane.go]: Just a language point check: is `dp.reportHealth` valid as soon as `dp` has been constructed?  (I'm not sure why it wouldn't be, but just being paranoid.)

File: dataplane/linux/int_dataplane.go [hunk 3]
```diff
@@ -1077,6 +1081,7 @@ func (d *InternalDataplane) apply() {
 		ipSetsWG.Add(1)
 		go func(ipSets *ipsets.IPSets) {
 			ipSets.ApplyUpdates()
+			d.reportHealth()
```
Comment (neiljerram) [inline dataplane/linux/int_dataplane.go]: Will this be thread-safe?
At this level:
```
	if d.config.HealthAggregator != nil {
		d.config.HealthAggregator.Report(
			healthName,
			&health.HealthReport{Live: true, Ready: d.doneFirstApply},
		)
	}
```
`d.config.HealthAggregator` should be OK because only set at start of day.  But maybe `d.doneFirstApply` could be changing in another goroutine, and we might need to delve further inside `d.config.HealthAggregator.Report`.

Comment (neiljerram) [reply dataplane/linux/int_dataplane.go]: Similarly the next 3 changes are all in goroutines.

File: calc/async_calc_graph.go (hunk 1 above)
Comment (fasaxc) [reply calc/async_calc_graph.go]: Thinking was that it might cause a lot of log spam from the health aggregator?

File: dataplane/linux/int_dataplane.go (hunk 3 above)
Comment (fasaxc) [reply dataplane/linux/int_dataplane.go]: I think it's correct but I wish it was a bit simpler:

- `d.config.HealthAggregator` is start-of day config
- `doneFirstApply` only changes on the main goroutine and the goroutines from which I'm calling `reportHealth()` are short-lived: the main goroutine starts them and then waits for them to finish before continuing.
- `Report()` itself uses a mutex to protect the HealthAggregator.

File: dataplane/linux/int_dataplane.go (hunk 2 above)
Comment (fasaxc) [reply dataplane/linux/int_dataplane.go]: Yes, I think that's fine in go.

File: calc/async_calc_graph.go (hunk 1 above)
Comment (neiljerram) [reply calc/async_calc_graph.go]: I think the aggregator report-time logging is all debug, and it might be better if needed to handle any log rate limiting inside the aggregator?  I think I'd prefer to err initially on the side of over-reporting from Felix.

File: dataplane/linux/int_dataplane.go (hunk 3 above)
Comment (neiljerram) [reply dataplane/linux/int_dataplane.go]: OK, thanks for the reassurance.

File: dataplane/linux/int_dataplane.go (hunk 2 above)
Comment (neiljerram) [reply dataplane/linux/int_dataplane.go]: Thanks.

File: calc/async_calc_graph.go (hunk 1 above)
Comment (fasaxc) [reply calc/async_calc_graph.go]: Benchmarked the report function at 500ns; that's probably OK.  Removing the logCtx from the report function knocks it down to 90ns in case we need to optimise that in future.

**Comment by fasaxc** [comment]
Should we also simply bump up the liveness timeout?

File: calc/async_calc_graph.go [hunk 4]
```diff
@@ -141,14 +142,18 @@ func (acg *AsyncCalcGraph) loop() {
 			case []api.Update:
 				// Update; send it to the dispatcher.
 				log.Debug("Pulled []KVPair off channel")
-				updStartTime := time.Now()
-				acg.AllUpdDispatcher.OnUpdates(update)
-				summaryUpdateTime.Observe(time.Since(updStartTime).Seconds())
-				// Record stats for the number of messages processed.
-				for _, upd := range update {
+				for i, upd := range update {
```
Comment (fasaxc) [inline calc/async_calc_graph.go]: @neiljerram I realised I had put the call to reportHealth() in the wrong place.  Needed to split up the calls to OnUpdates instead.
